lib/internal/Magento/Framework/CompiledInterception/Generator/CompiledInterceptor.php

```diff
@@ -0,0 +1,399 @@
+            'body' => implode("\n", $body),
+            'docblock' => ['shortDescription' => '{@inheritdoc}'],
+        ];
+
+    }
+
+    protected function _getClassMethods()
+    {
+        return $this->classMethods;
+    }
+
+    protected function _getClassProperties()
+    {
+        return $this->classProperties;
+    }
+
+    private function addCodeSubBlock(&$body, $sub, $indent = 1)
+    {
+        foreach ($sub as $line) {
+            $body[] = str_repeat("\t", $indent) . $line;
+        }
+    }
+
+    /**
+     * @param \ReflectionMethod $method
+     * @param $conf
+     * @param $parameters
+     * @return array
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     * @SuppressWarnings(PHPMD.NPathComplexity)
+     */
+    protected function _getMethodSourceFromConfig(\ReflectionMethod $method, $conf, $parameters, $returnVoid)
+    {
+        $body = [];
+        $first = true;
+        $capName = ucfirst($method->getName());
+        $extraParams = empty($parameters) ? '' : (', ' . $this->_getParameterList($parameters));
+
+        if (isset($conf[DefinitionInterface::LISTENER_BEFORE])) {
+            foreach ($conf[DefinitionInterface::LISTENER_BEFORE] as $plugin) {
+                if ($first) $first = false; else $body[] = "";
+
+                $call = "\$this->" . $this->getGetterName($plugin) . "()->before$capName(\$this$extraParams);";
+
+                if (!empty($parameters)) {
+                    $body[] = "\$beforeResult = " . $call;
+                    $body[] = "if (\$beforeResult !== null) list({$this->_getParameterList($parameters)}) = (array)\$beforeResult;";
+                } else {
+                    $body[] = $call;
+                }
+            }
+        }
+
+
+        $chain = [];
+        $main = [];
+        if (isset($conf[DefinitionInterface::LISTENER_AROUND])) {
+            $plugin = $conf[DefinitionInterface::LISTENER_AROUND];
+            $main[] = "\$this->" . $this->getGetterName($plugin) . "()->around$capName(\$this, function({$this->_getParameterListForNextCallback($parameters)}){";
+            $this->addCodeSubBlock($main, $this->_getMethodSourceFromConfig($method, $plugin['next'] ?: [], $parameters, $returnVoid));
+            $main[] = "}$extraParams);";
+        } else {
+            $main[] = "parent::{$method->getName()}({$this->_getParameterList($parameters)});";
+        }
+        $chain[] = $main;
+
+        if (isset($conf[DefinitionInterface::LISTENER_AFTER])) {
+            foreach ($conf[DefinitionInterface::LISTENER_AFTER] as $plugin) {
+                if ($returnVoid) {
+                    $chain[] = ["((\$tmp = \$this->" . $this->getGetterName($plugin) . "()->after$capName(\$this, \$result$extraParams)) !== null) ? \$tmp : \$result;"];
+                } else {
+                    $chain[] = ["\$this->" . $this->getGetterName($plugin) . "()->after$capName(\$this, \$result$extraParams);"];
+                }
+            }
+        }
```

Comment:
we can introduce either 3 corresponding methods or classes which will generate listeners before/around/after

avoiding introducing too many conditional logic to the current method.
Doing so we will make this method much easier to read , and remove annotation for 
```
     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
     * @SuppressWarnings(PHPMD.NPathComplexity)
```

Reply:
this was a good idea. fixed